### PR TITLE
Resource tagging in the registry-creds up command

### PR DESCRIPTION
**Pull request by PettitWesley**
*Description of changes:*
Tag SM secrets and new IAM roles. (IAM is not supported by the standalone resource tagging API, so tagging can only be supported during role creation).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by PettitWesley** [comment]
Testing:

```
$ ecs-cli registry-creds up reg-creds.yml --role-name cat-role2 --tags impala=tame,parker=kevin
INFO[0000] Processing credentials for registry my-registry.example.com...
INFO[0000] Existing credential secret found, using arn:aws:secretsmanager:us-west-2:144718711470:secret:amazon-ecs-cli-setup-my-registry.example.com-VUHL4v
INFO[0000] Creating resources for task execution role cat-role2...
INFO[0000] Created new task execution role arn:aws:iam::144718711470:role/cat-role2
INFO[0000] Created new task execution role policy arn:aws:iam::144718711470:policy/amazon-ecs-cli-setup-cat-role2-policy-20190129T005157Z
INFO[0000] Attached AWS managed policy arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy to role cat-role2
INFO[0000] Attached new policy arn:aws:iam::144718711470:policy/amazon-ecs-cli-setup-cat-role2-policy-20190129T005157Z to role cat-role2
INFO[0001] Writing registry credential output to new file /Users/wppttt/stuffnthings/bugbash/compose3/ecs-registry-creds_20190129T005157Z.yml
INFO[0001]
If your input file contains sensitive information, make sure that you delete it after use.
```

```
$ aws resourcegroupstaggingapi get-resources --tag-filters Key=impala,Values=tame --region us-west-2
{
    "ResourceTagMappingList": [
        {
            "ResourceARN": "arn:aws:secretsmanager:us-west-2:144718711470:secret:amazon-ecs-cli-setup-my-registry.example.com-VUHL4v",
            "Tags": [
                {
                    "Key": "impala",
                    "Value": "tame"
                },
                {
                    "Key": "parker",
                    "Value": "kevin"
                }
            ]
        }
    ]
}
```

```
aws iam list-role-tags --role-name cat-role2
{
    "Tags": [
        {
            "Key": "impala",
            "Value": "tame"
        },
        {
            "Key": "parker",
            "Value": "kevin"
        }
    ],
    "IsTruncated": false
}
```

File: ecs-cli/modules/cli/regcreds/regcreds_app.go [hunk 1]
```diff
@@ -21,10 +21,13 @@ import (
 	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients/aws/iam"
 	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients/aws/kms"
 	secretsClient "github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients/aws/secretsmanager"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients/aws/tagging"
 	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/commands/flags"
 	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/config"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/utils"
 	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/utils/regcredio"
 	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi"
```
Comment (allisaurus) [inline ecs-cli/modules/cli/regcreds/regcreds_app.go]: can we maybe alias this as "taggingSdk" ?

File: ecs-cli/modules/cli/regcreds/create_task_execution_role.go [hunk 2]
```diff
@@ -119,3 +120,15 @@ func attachRolePolicies(secretPolicyARN, roleName, region string, client iamClie
 
 	return nil
 }
+
+func convertToIAMTags(tags map[string]*string) []*iam.Tag {
+	var iamTags []*iam.Tag
+	for key, value := range tags {
+		iamTags = append(iamTags, &iam.Tag{
+			Key:   aws.String(key),
+			Value: value,
```
Comment (allisaurus) [inline ecs-cli/modules/cli/regcreds/create_task_execution_role.go]: Is it valid to have a `nil` value for a tag? If so, can we add a case for it to `TestCreateTaskExecutionRoleWithTags` ?

**Comment by PettitWesley** [comment]
@allisaurus 

```
$ ecs-cli registry-creds up reg-creds.yml --role-name cat-role2 --update-existing-secrets --tags InnerSpeaker=favoritealbum --region us-west-2
INFO[0000] Processing credentials for registry my-registry.example.com...
INFO[0000] Existing credential secret found, using arn:aws:secretsmanager:us-west-2:144718711470:secret:amazon-ecs-cli-setup-my-registry.example.com-VUHL4v
INFO[0000] Creating resources for task execution role cat-role2...
INFO[0000] Using existing role cat-role2
INFO[0000] Created new task execution role policy arn:aws:iam::144718711470:policy/amazon-ecs-cli-setup-cat-role2-policy-20190205T025752Z
INFO[0000] Attached AWS managed policy arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy to role cat-role2
INFO[0000] Attached new policy arn:aws:iam::144718711470:policy/amazon-ecs-cli-setup-cat-role2-policy-20190205T025752Z to role cat-role2
INFO[0001] Writing registry credential output to new file /Users/wppttt/stuffnthings/bugbash/compose3/ecs-registry-creds_20190205T025752Z.yml
INFO[0001]
If your input file contains sensitive information, make sure that you delete it after use.
```

New and old tags are present on the secret:
```
$ aws resourcegroupstaggingapi get-resources --tag-filters Key=InnerSpeaker,Values=favoritealbum --region us-west-2
{
    "ResourceTagMappingList": [
        {
            "ResourceARN": "arn:aws:secretsmanager:us-west-2:144718711470:secret:amazon-ecs-cli-setup-my-registry.example.com-VUHL4v",
            "Tags": [
                {
                    "Key": "InnerSpeaker",
                    "Value": "favoritealbum"
                },
                {
                    "Key": "impala",
                    "Value": "tame"
                },
                {
                    "Key": "parker",
                    "Value": "kevin"
                }
            ]
        }
    ]
}
```

File: ecs-cli/modules/cli/regcreds/create_task_execution_role.go (hunk 2 above)
Comment (PettitWesley) [reply ecs-cli/modules/cli/regcreds/create_task_execution_role.go]: Docs say its a required field: https://docs.aws.amazon.com/sdk-for-go/api/service/iam/#Tag

You can pass an empty string though, which my code handles correctly:

```
$ ecs-cli registry-creds up --role-name catsss --tags cat=,meow= reg-creds.yml
INFO[0000] Processing credentials for registry my-registry.example.com...
INFO[0000] Existing credential secret found, using arn:aws:secretsmanager:us-west-1:REDACTED:secret:amazon-ecs-cli-setup-my-registry.example.com-TIxAwp
INFO[0000] Creating resources for task execution role catsss...
INFO[0000] Created new task execution role arn:aws:iam::REDACTED:role/catsss
INFO[0000] Created new task execution role policy arn:aws:iam::REDACTED:policy/amazon-ecs-cli-setup-catsss-policy-20190213T184615Z
INFO[0001] Attached AWS managed policy arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy to role catsss
INFO[0001] Attached new policy arn:aws:iam::REDACTED:policy/amazon-ecs-cli-setup-catsss-policy-20190213T184615Z to role catsss
INFO[0001] Writing registry credential output to new file stuffnthings/bugbash/compose3/ecs-registry-creds_20190213T184615Z.yml
INFO[0001]
If your input file contains sensitive information, make sure that you delete it after use.
```

(I checked that the tags had been set in the Console)

File: ecs-cli/modules/cli/regcreds/regcreds_app.go [hunk 3]
```diff
@@ -21,10 +21,13 @@ import (
 	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients/aws/iam"
 	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients/aws/kms"
 	secretsClient "github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients/aws/secretsmanager"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/clients/aws/tagging"
 	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/commands/flags"
 	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/config"
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/utils"
 	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/utils/regcredio"
 	"github.com/aws/aws-sdk-go/aws"
+	taggngSDK "github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi"
```
Comment (allisaurus) [inline ecs-cli/modules/cli/regcreds/regcreds_app.go]: nit: should be `taggingSDK`

File: ecs-cli/modules/cli/regcreds/regcreds_app.go [hunk 4]
```diff
@@ -306,3 +326,26 @@ func validateOutputOptions(outputDir string, skipOutput bool) error {
...
+		ResourceARNList: arns,
+		Tags:            tags,
+	}
+	output, err := taggingClient.TagResources(input)
+	if err != nil {
+		return err
+	}
+
+	for resource, info := range output.FailedResourcesMap {
+		return fmt.Errorf("Failed to tag resource %s; error=%s", resource, *info.ErrorMessage)
```
Comment (allisaurus) [inline ecs-cli/modules/cli/regcreds/regcreds_app.go]: can we add test for this error case (FailedResourceMap != nil) and the API error return (line 342)?

File: ecs-cli/modules/commands/regcreds/regcreds_command.go [hunk 5]
```diff
@@ -65,5 +65,9 @@ func regcredsUpFlags() []cli.Flag {
 			Name:  flags.OutputDirFlag,
 			Usage: "[Optional] The directory where the output file should be created. If none specified, file will be created in the current working directory.",
 		},
+		cli.StringFlag{
+			Name:  flags.ResourceTagsFlag,
+			Usage: "[Optional] Add AWS Resource tags to the Secrets Manager secrets and IAM Role. Note that tags can only be added to the IAM role at time of creation.",
```
Comment (allisaurus) [inline ecs-cli/modules/commands/regcreds/regcreds_command.go]: Suggested edit for conciseness + consistency w/ other flags: 
"The AWS Resource tags to add to the Secrets Manager secrets and new IAM Role. Existing IAM Roles cannot be tagged."

File: ecs-cli/modules/clients/aws/iam/client.go [hunk 6]
```diff
@@ -81,12 +81,15 @@ func (c *iamClient) CreatePolicy(input iam.CreatePolicyInput) (*iam.CreatePolicy
...
 		AssumeRolePolicyDocument: aws.String(assumeRolePolicyDoc),
 		Description:              aws.String(roleDescription),
 		RoleName:                 aws.String(roleName),
 	}
+	if len(tags) > 0 {
+		createRoleRequest.Tags = tags
+	}
 	roleResult, err := c.CreateRole(createRoleRequest)
 	// if err is b/c role already exists, OK to continue
 	if err != nil && !utils.EntityAlreadyExists(err) {
```
Comment (allisaurus) [inline ecs-cli/modules/clients/aws/iam/client.go]: What happens if we try to apply tags to an existing role? Does it return a distinct error or succeed? If the former, wondering if we should be checking for a specific error here & allowing it to succeed (like with `EntityAlreadyExists`) or let it fail.

Comment (PettitWesley) [reply ecs-cli/modules/clients/aws/iam/client.go]: Adding here what we discussed: trying to re-create an existing role but with tags will still fail with an `EntityAlreadyExists` error.

**Comment by allisaurus** [comment]
LGTM! Though do consider the suggested flag text change.

File: ecs-cli/modules/clients/aws/iam/client.go (hunk 6 above)
Comment (SoManyHs) [reply ecs-cli/modules/clients/aws/iam/client.go]: I'm happy to go along with what @allisaurus agrees with but curious why this is the correct behavior?

File: ecs-cli/modules/cli/regcreds/create_task_execution_role.go (hunk 2 above)
Comment (SoManyHs) [reply ecs-cli/modules/cli/regcreds/create_task_execution_role.go]: Do we want to log info about what tags are being applied to what resources?

File: ecs-cli/modules/clients/aws/iam/client.go (hunk 6 above)
Comment (SoManyHs) [reply ecs-cli/modules/clients/aws/iam/client.go]: As per offline discussion, I understand now that this is strictly on creation of a new role, so this seems fine.
